### PR TITLE
Refactor progress indicators

### PR DESCRIPTION
**Pull request by xbrianh**
- Add progress reporter
- Rate limit output to keep things perty
- Add tests

### CODE REVIEW

File: terra_notebook_utils/progress.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        self.steps_completed += number_of_steps
+        slots_filled = int(self._length * self.steps_completed / self.number_of_steps)
+        duration = ceil(time.time() - self._start_time)
+        portion_complete = self.steps_completed / self.number_of_steps
+        bar = "#" * slots_filled
+        bar += " " * (self._length - slots_filled)
+        bar += f" {floor(100 * portion_complete)}%"
+        if self.size is not None:
+            bar += f" of {self.size}{self.units}"
+            bar += f" %.2f{self.units}/s" % (self.size * portion_complete / duration)
```
Comment (DailyDreaming) [inline terra_notebook_utils/progress.py]: Nit: Mixing f-strings and %-formatting.  XD

File: terra_notebook_utils/progress.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        self._last_call = time.time() - 10 * self.period
+
+    def __call__(self, func):
+        def wrapper(*args, **kwargs):
+            now = time.time()
+            if now - self._last_call >= self.period:
+                self._last_call = now
+                return func(*args, **kwargs)
+            elif self.exception is not None:
+                raise self.exception("Too soon")
```
Comment (DailyDreaming) [inline terra_notebook_utils/progress.py]: lol

File: terra_notebook_utils/progress.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    def checkpoint(self, _units_processed: int):
+        self._units_processed = [self._units_processed[-1], self._units_processed[-1] + _units_processed]
+        self._checkpoint_time = [self._checkpoint_time[-1], time.time()]
+        self._print()
+
+    @RateLimited(2.0)
+    def _print(self):
+        with self._lock:
+            print("\r",
+                  f"%9i {self.units} processed," % self._units_processed[-1],
```
Comment (DailyDreaming) [inline terra_notebook_utils/progress.py]: Same formatting nit here and below.

File: terra_notebook_utils/progress.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+
+class ProgressBar(AbstractContextManager):
+    def __init__(self, number_of_steps, prefix="", units="", size: int=None):
+        self.number_of_steps = number_of_steps
+        self.steps_completed = 0
+        self.prefix = prefix
+        self.units = units
+        self.size = size
+        self._length = 40
```
Comment (DailyDreaming) [inline terra_notebook_utils/progress.py]: Seems like a magic number?  Maybe a comment?

File: terra_notebook_utils/progress.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    @property
+    def duration(self):
+        return self._checkpoint_time[-1] - self.start_time
+
+    @property
+    def average_rate(self):
+        return self._units_processed[-1] / self.duration
+
+    @property
+    def instantaneous_rate(self):
```
Comment (DailyDreaming) [inline terra_notebook_utils/progress.py]: Nit: is this function called?

File: terra_notebook_utils/progress.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    Decorator to rate limit method calls. Raises `exception` if method is called more quickly than
+    `rate` times per second, or does nothing if exception is None.
+    """
+    def __init__(self, rate: float, exception: Exception=None):
+        self.period = 1 / rate
+        self.exception = exception
+        self.reset()
+
+    def reset(self):
+        self._last_call = time.time() - 10 * self.period
```
Comment (DailyDreaming) [inline terra_notebook_utils/progress.py]: What does the `10` account for?

File: tests/test_terra_notebook_utils.py [hunk 7]
```diff
@@ -86,5 +90,49 @@ def test_extract(self):
...
+            @progress.RateLimited(rate, MockRateLimitedException)
+            def raising_rate_limited_func():
+                pass
+
+            with self.assertRaises(MockRateLimitedException):
+                raising_rate_limited_func()
+                raising_rate_limited_func()
+
+        with self.subTest("Should be able to avoid exceptions using reset"):
+            raising_rate_limited_func.reset()
```
Comment (DailyDreaming) [inline tests/test_terra_notebook_utils.py]: Is it consistent if this is done in a `for loop` for many more times?

File: tests/test_terra_notebook_utils.py [hunk 8]
```diff
@@ -86,5 +90,49 @@ def test_extract(self):
...
+        with self.subTest("Should be rate limited to expected number of calls"):
+            call_info = dict(number_of_calls=0)
+
+            @progress.RateLimited(rate)
+            def rate_limited_func():
+                call_info['number_of_calls'] += 1
+
+            start_time = time.time()
+            duration = 0
+            while duration <= 4.5 * 1 / rate:
```
Comment (DailyDreaming) [inline tests/test_terra_notebook_utils.py]: I'd expect this to be `5 * 1`?

File: tests/test_terra_notebook_utils.py [hunk 9]
```diff
@@ -86,5 +90,49 @@ def test_extract(self):
...
+        with self.subTest("Should get expected exception for calling too quickly"):
+            class MockRateLimitedException(Exception):
+                pass
+
+            @progress.RateLimited(rate, MockRateLimitedException)
+            def raising_rate_limited_func():
+                pass
+
+            with self.assertRaises(MockRateLimitedException):
+                raising_rate_limited_func()
```
Comment (DailyDreaming) [inline tests/test_terra_notebook_utils.py]: Why does this call `raising_rate_limited_func()` twice?

File: terra_notebook_utils/progress.py (hunk 6 above)
Comment (xbrianh) [reply terra_notebook_utils/progress.py]: It's just an arbitrary factor to guarantee the function is not rate limited on the next call. anything larger than 1 should work.

File: terra_notebook_utils/progress.py (hunk 4 above)
Comment (xbrianh) [reply terra_notebook_utils/progress.py]: Not magic so much as a choice for a reasonable length. I'll add a comment.

File: terra_notebook_utils/progress.py (hunk 1 above)
Comment (xbrianh) [reply terra_notebook_utils/progress.py]: What's wrong with it?

File: terra_notebook_utils/progress.py (hunk 5 above)
Comment (xbrianh) [reply terra_notebook_utils/progress.py]: Nope, but it seems like it'll be occasionally useful.

File: tests/test_terra_notebook_utils.py (hunk 8 above)
Comment (xbrianh) [reply tests/test_terra_notebook_utils.py]: The first call is at `t=0`. If we waited for 5 seconds to ellapse I'd expect to get 6 calls, not 5.

File: tests/test_terra_notebook_utils.py (hunk 9 above)
Comment (xbrianh) [reply tests/test_terra_notebook_utils.py]: Calling it once wouldn't cause the rate limiting logic to trigger.

File: tests/test_terra_notebook_utils.py (hunk 7 above)
Comment (xbrianh) [reply tests/test_terra_notebook_utils.py]: That would work, but wouldn't test any additional logic.

Switched to loops to improve aesthetics.

**Pull request merged**